python/lsst/daf/butler/core/queries/query.py

```diff
@@ -0,0 +1,229 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+from __future__ import annotations
+
+__all__ = ("Query",)
+
+import itertools
+from typing import Optional, Dict, Any, Tuple, Callable
+
+from sqlalchemy.sql import FromClause
+from sqlalchemy.engine import RowProxy, ResultProxy, Connection
+
+from lsst.sphgeom import Region
+
+from ..datasets import DatasetType, DatasetRef
+from ..dimensions import DimensionGraph, DataCoordinate, ExpandedDataCoordinate
+
+from .structs import QuerySummary, QueryColumns, QueryParameters
+
+
+class Query:
+    """A wrapper for a SQLAlchemy query that knows how to re-bind parameters
+    and transform result rows into data IDs and dataset references.
+
+    A `Query` should almost always be constructed directly by a call to
+    `QueryBuilder` finish; direct construction will make it difficult to be
+    able to maintain invariants between arguments (see the documentation for
+    `QueryColumns` and `QueryParameters` for more information).
+
+    Parameters
+    ----------
+    connection: `sqlalchemy.engine.Connection`
+        Connection used to execute the query.
+    sql : `sqlalchemy.sql.FromClause`
+        A complete SELECT query, including at least SELECT, FROM, and WHERE
+        clauses.
+    summary : `QuerySummary`
+        Struct that organizes the dimensions involved in the query.
+    columns : `QueryColumns`
+        Columns that are referenced in the query in any clause.
+    parameters : `QueryParameters`
+        Bind parameters for the query.
+
+    Notes
+    -----
+    SQLAlchemy is used in the public interface of `Query` rather than just its
+    implementation simply because avoiding this would entail writing wrappers
+    for the `sqlalchemy.engine.RowProxy` and `sqlalchemy.engine.ResultProxy`
+    classes that are themselves generic wrappers for lower-level Python DBAPI
+    classes.  Another layer would entail another set of computational
+    overheads, but the only reason we would seriously consider not using
+    SQLAlchemy here in the future would be to reduce computational overheads.
+    """
+
+    def __init__(self, *, connection: Connection, sql: FromClause,
+                 summary: QuerySummary, columns: QueryColumns, parameters: QueryParameters):
+        self.summary = summary
+        self.sql = sql
+        self._columns = columns
+        self._parameters = parameters
+        self._connection = connection
+
+    def predicate(self, region: Optional[Region] = None) -> Callable[[RowProxy], bool]:
+        """Return a callable that can perform extra Python-side filtering of
+        query results.
+
+        To get the expected results from a query, the returned predicate *must*
+        be used to ignore rows for which it returns `False`; this permits the
+        `QueryBuilder` implementation to move logic from the database to Python
+        without changing the public interface.
+
```

Comment:
:+1: on docs.

By returning a new callable, we can do some work up-front instead of per-row.  We're not really using that much right now, but I'd like to keep the interface this way in case that becomes valuable in the future.